cmd/client_test.go

```diff
@@ -0,0 +1,213 @@
+
+	cloudprober.Start(testHermes.Ctx)
+
+	client, errClient := NewClient(rpcServer)
+	if errClient != nil {
+		t.Fatalf("Cloudprober gRPC Client could not be initialised: %v", errClient)
+	}
+
+	return testHermes, client
+}
+
+func teardownCloudproberAndClient(t *testing.T, testHermes *hermes.Hermes, client *CloudproberClient) {
+	// TODO(evanSpendlove): Find a safe way to close down Cloudprober and restart it.
+	// Check if the probe was added correctly
+	t.Helper()
+	listProbesResp, err := client.ListProbes(testHermes.Ctx)
+	if err != nil {
+		t.Errorf("ListProbes() failed, expected error: %v, got: %v\n\n", nil, err)
+	}
+
+	probesList := listProbesResp.GetProbe()
+
+	for _, probe := range probesList {
+		if err := client.RemoveProbe(probe.GetName(), testHermes.Ctx); err != nil {
+			t.Errorf("RemoveProbe() failed, expected error: %v, got %v", nil, err)
+		}
+	}
+
+	client.CloseConn()
+}
+
+// TODO(evanSpendlove): Replace this test with a test using a Hermes probe definition
+// generationRedisProbeDef generates an inline probe definition of the RedisProbe probe extension.
+// The RedisProbe extension is supplied as an example extension in Cloudprober.
+func generateRedisProbeDef(name string) *probes_configpb.ProbeDef {
+	probeDef := &probes_configpb.ProbeDef{ // Create probe def
+		Name: proto.String(name),
+		Type: probes_configpb.ProbeDef_EXTENSION.Enum(),
+		Targets: &targetspb.TargetsDef{
+			Type: &targetspb.TargetsDef_DummyTargets{},
+		},
+	}
+
+	// Add RedisProbe extension to probeDef proto
+	op := myprobe.ProbeConf_Op.Enum(myprobe.ProbeConf_SET)
+	proto.SetExtension(probeDef, myprobe.E_RedisProbe, &myprobe.ProbeConf{Op: op, Key: proto.String("testkey"), Value: proto.String("testval")})
+	return probeDef
+}
+
+// TestRegisterAndAddProbe tests that an extension probe type can be registered and added to Cloudprober without error.
+func TestRegisterAndAddProbe(t *testing.T) {
+	testHermes, client := setupCloudproberAndClient(t)
+	defer teardownCloudproberAndClient(t, testHermes, client)
+
+	for i := 0; i < testRunCount; i++ {
+		probe := fmt.Sprintf("%s%d", "testExtension", i)
+		if err := client.RegisterAndAddProbe(int(myprobe.E_RedisProbe.Field), testHermes.Ctx, generateRedisProbeDef(probe), &myprobe.Probe{}); err != nil {
+			t.Fatalf("Probe not correctly registered and added to Cloudprober, expected error: %v, got: %v", nil, err)
+		}
+
+		// Check if the probe was added correctly
+		listProbesResp, err := client.ListProbes(testHermes.Ctx)
+		if err != nil {
+			t.Fatalf("ListProbes() failed, expected error: %v, got: %v", nil, err)
+		}
+
+		probesList := listProbesResp.GetProbe()
+
+		// Sorts the probes by their extension number as they are in the format:
+		// "testExtension" + <number>, e.g. "testExtension0".
+		sort.SliceStable(probesList[:], func(i, j int) bool {
+			probeNum0, _ := strconv.Atoi(probesList[i].GetName()[13:])
+			probeNum1, _ := strconv.Atoi(probesList[j].GetName()[13:])
+			return probeNum0 < probeNum1
```

Comment:
Maybe we should make ListProbes return an already sorted output so you don't have to worry about sorting these in the tests?

Reply:
I've made this change, but it is worth noting that it does alter the return value so that it is a sorted array of `*Probe` objects, as there is no way to set this array inside the `ListProbesResponse{}` struct.